src/NosCore.Controllers/Inventory/UpgradePacketHandler.cs

```diff
@@ -0,0 +1,41 @@
+//  __  _  __    __   ___ __  ___ ___
+// |  \| |/__\ /' _/ / _//__\| _ \ __|
+// | | ' | \/ |`._`.| \_| \/ | v / _|
+// |_|\__|\__/ |___/ \__/\__/|_|_\___|
+// 
+// Copyright (C) 2019 - NosCore
+// 
+// NosCore is a free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or any later version.
+// 
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+// 
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+using ChickenAPI.Packets.ClientPackets.Player;
+using NosCore.GameObject;
+using NosCore.GameObject.Networking.ClientSession;
+using NosCore.GameObject.Providers.UpgradeProvider;
+
+namespace NosCore.PacketHandlers.Inventory
+{
+    public class UpgradePacketHandler : PacketHandler<UpgradePacket>, IWorldPacketHandler
+    {
+        private readonly IUpgradeProvider _upgradeService;
+
+        public UpgradePacketHandler(IUpgradeProvider upgradeService)
+        {
+            _upgradeService = upgradeService;
+        }
+
+        public override void Execute(UpgradePacket packet, ClientSession clientSession)
+        {
+            _upgradeService.UpgradeLaunch(clientSession, packet);
```

Comment:
I wonder if it wouldn’t be better to have item removal here and just to make the service responsible of merge

Reply:
Hum, I don't know, if you wan't i can do it, but is it usefull ?

Reply:
Will we use service one day without remove items cost from inventory ?

Reply:
Other type of upgrade.